### PR TITLE
Add limited PR testing

### PR DESCRIPTION
**Pull request by cwpearson**
Add PR tests a variety of exercises against Kokkos / Kokkos Kernels `master` branches. This only builds the exercises, it does not run their solutions.

Since it's a bit clumsy to build every exercise, this CI uses a script to do it (batch for Windows, bash for Linux/macOS).

- [x] Linux / g++ / OpenMP
- [x] macOS / clang++ / Threads
- [x] Windows / cl / Serial 
- [x] Linux / nvcc / CUDA
- [x] More exercises to test?
- [ ] Run Solutions where possible?

Many exercises did not immediately build, have certain execution spaces hardcoded, or had other problems. Resolving those is left as future work.

### CODE REVIEW

**Review by cedricchevalier19 — COMMENTED** [review]
I think it is a very good start.

In the future we can remove the scripts (with a global CMake #91). Perhaps these scripts can be relocated in a more hidden place?

The nightly should be run from Kokkos repo using workflow dispatch.

**Review by cedricchevalier19 — APPROVED** [review]
I think we cannot run the proposed action while `.github/workflows` is empty.
@cwpearson, did you run the PR with a runner like https://github.com/nektos/act ?

**Comment by masterleinad** [comment]
> @cwpearson, did you run the PR with a runner like [nektos/act](https://github.com/nektos/act) ?

See https://github.com/cwpearson/kokkos-tutorials/actions/runs/11883293720, https://github.com/cwpearson/kokkos-tutorials/actions/runs/11883293721, and https://github.com/cwpearson/kokkos-tutorials/actions/runs/11883293723.

**Review by masterleinad — APPROVED** [review]
Looks good as a first step.

File: .github/workflows/github-pr-linux-container.yaml [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+    - name: Configure Kokkos
+      run: >
+        cmake -S "${GITHUB_WORKSPACE}"/kokkos -B "${GITHUB_WORKSPACE}"/build-kokkos
+        -DCMAKE_INSTALL_PREFIX="${GITHUB_WORKSPACE}"/install-kokkos
+        -DCMAKE_CXX_COMPILER="${GITHUB_WORKSPACE}"/kokkos/bin/nvcc_wrapper
+        -DCMAKE_BUILD_TYPE=RelWithDebInfo
+        -DKokkos_ENABLE_CUDA=ON
+        -DKokkos_ARCH_AMPERE80=ON
+        -DKokkos_ENABLE_COMPILER_WARNINGS=ON
```
Comment (dalg24) [inline .github/workflows/github-pr-linux-container.yaml:47]:
```suggestion
```

File: .github/workflows/github-pr-linux-container.yaml [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        repository: 'kokkos/kokkos-kernels'
+        ref: master
+        path: kokkos-kernels
+
+    - name: Configure Kokkos
+      run: >
+        cmake -S "${GITHUB_WORKSPACE}"/kokkos -B "${GITHUB_WORKSPACE}"/build-kokkos
+        -DCMAKE_INSTALL_PREFIX="${GITHUB_WORKSPACE}"/install-kokkos
+        -DCMAKE_CXX_COMPILER="${GITHUB_WORKSPACE}"/kokkos/bin/nvcc_wrapper
+        -DCMAKE_BUILD_TYPE=RelWithDebInfo
```
Comment (dalg24) [inline .github/workflows/github-pr-linux-container.yaml:44]: Fine but noting that it is the default hence unnecessary

File: .github/workflows/github-pr-windows.yaml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    - name: Checkout Kokkos
+      uses: actions/checkout@v4
+      with:
+        repository: 'kokkos/kokkos'
+        ref: master
+        path: kokkos
+
+    - name: Configure Kokkos
+      run: >
+        cmake -S ${{ github.workspace}}\kokkos -B ${{ github.workspace}}\build-kokkos
```
Comment (dalg24) [inline .github/workflows/github-pr-windows.yaml:37]: Did you mean these backslashes instead of regular slashes?

File: Scripts/ci-configure-build-test.bat
Comment (dalg24) [inline Scripts/ci-configure-build-test.bat:1]: Where/how is that .bat file used?

File: Scripts/ci-configure-build-test.sh [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  EXERCISES+=(mpi_heat_conduction/Solution) # TODO: mpi_heat_conduction/Begin does not use cmake
+fi
+
+# Add Begin and Solution subdirectory from BEGIN_SOLUTION_EXCERCISES to EXERCISES
+for e in "${BEGIN_SOLUTION_EXERCISES[@]}"; do
+  EXERCISES+=("$e"/Begin)
+  EXERCISES+=("$e"/Solution)
+done
+
+export Kokkos_ROOT="$kokkos_root"
```
Comment (dalg24) [inline Scripts/ci-configure-build-test.sh:83]: Why do we handle Kokkos and KokkosKernels differently and why don't we just set these environment variable externally?

File: Scripts/ci-configure-build-test.sh [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+kokkos_root="$1"
+kernels_root="$2"
+tutorials_src="$3"
+cpp_compiler="$4"
+build_type="$5"
+backend="$6"
+
+# These are exercises with CMakeLists.txt in Begin and Solution subdirectories
+# TODO: advanced_reductions seems broken
```
Comment (masterleinad) [inline Scripts/ci-configure-build-test.sh:13]: Can you elaborate?

File: Scripts/ci-configure-build-test.sh [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+# TODO: hpcbind does not use cmake
+# TODO: instances does not use cmake
+# TODO: parallel_scan seems broken
```
Comment (masterleinad) [inline Scripts/ci-configure-build-test.sh:16]: Can you elaborate?

File: Scripts/ci-configure-build-test.sh (hunk 5 above)
Comment (masterleinad) [reply Scripts/ci-configure-build-test.sh:13]: Ah, `Begin` doesn't compile.

File: .github/workflows/github-pr-windows.yaml (hunk 3 above)
Comment (cwpearson) [reply .github/workflows/github-pr-windows.yaml:37]: This is the Windows build, this is how I would have written it long ago, can I just use backslashes on Windows now?

File: Scripts/ci-configure-build-test.bat
Comment (cwpearson) [reply Scripts/ci-configure-build-test.bat:1]: It's used to drive the compilation and testing in the Windows build

File: Scripts/ci-configure-build-test.sh (hunk 6 above)
Comment (cwpearson) [reply Scripts/ci-configure-build-test.sh:16]: It just means there were compilation errors when I tried to compile it in this framework. I didn't spend more than  30s looking at any of them. I figured fixing them up could be follow-on work.

Comment (masterleinad) [reply Scripts/ci-configure-build-test.sh:16]: Yes, see #104.

**Comment by cwpearson** [comment]
Is this okay to merge for now, and we can clean up more later?

**Pull request merged**